Azure-Samples/azure-search-openai-demo
# Port to Quart

## Purpose

This PR ports the app to Quart, a Python ASGI web framework. 
Flask is a WSGI framework, which is synchronous. That means that when a request comes in, the server will wait for the request to be processed before it can handle another request. We have added support for multiple gunicorn workers/threads, but they can still easily be blocked by a long-running request. Each /chat request makes 4 API calls, which take from 0.2-7 seconds each.

Quart is designed with the same API as Flask, with a few changes needed for asynchronous support. Most of my changes are the addition of await/async. For the Azure SDK calls, I used the async versions of the clients (imported from `.aio`). For the OpenAI API calls, I used the async versions of the functions (`acreate` instead of `create`).

For deployment on App Service, I use gunicorn with the uvicorn worker.

## Does this introduce a breaking change?

```
[X] Yes
[ ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?


```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[X] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test

* `app/start.sh` should start the app
* `azd up` should deploy the app
* `python3 -m pytest` should run the tests
